Confluence PDF Download

WildDogOne opened the pull request:
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Description
The changes in this pull request facilitate the ability to download Confluence Pages as PDF documents.
This might help XSOAR users to automatically transfer vital documentation from a Confluence Server to an offsite location.
I also added Personal Access Token authentication since that was missing, however I am absolutely OK with splitting that into another pull requests since it has nothing directly to do with the PDF download.

## Must have
- [x] Tests -> There are no tests in the original confluence integration
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=39794) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=39794) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=39794) it.</sub>

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -239,6 +306,16 @@ def get_content_command():
+    pageid = demisto.args().get("pageid")
+    pdf = get_pdf(pageid)
+
+    demisto.results(fileResult(f"PDF for {pageid}", pdf))
```
```suggestion
    demisto.results(fileResult(f"Confluence_page_{page_id}.pdf", pdf))
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -205,6 +261,15 @@ def get_content(key, title):
     return res
 
 
+def get_pdf(pageid):
```
```suggestion
def get_pdf(page_id):
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -205,6 +261,15 @@ def get_content(key, title):
 
 
+def get_pdf(pageid):
+    params = {"pageId": pageid}
```
```suggestion
    params = {"pageId": page_id}
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.yml:
```diff
@@ -138,6 +143,43 @@ script:
     description: Returns Confluence content by space key and title.
+  - name: confluence-get-page-as-pdf
+    arguments:
+    - name: pageid
```
```suggestion
    - name: page_id
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.yml:
```diff
@@ -16,6 +16,11 @@ configuration:
   type: 9
   required: true
+- display: Personal Access Token
+  name: pat
+  defaultvalue: "false"
+  type: 8
```
```suggestion
  name: personal_access_token
  type: 9
  hiddenusername: true  
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.yml:
```diff
@@ -16,6 +16,11 @@ configuration:
   defaultvalue: ""
   type: 9
   required: true
```
```suggestion
  required: false
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
 USERNAME = demisto.params()["credentials"]["identifier"]
 PASSWORD = demisto.params()["credentials"]["password"]
-HEADERS = {"Content-Type": "application/json", "Accept": "application/json"}
+PAT = demisto.params().get("pat")
```
```suggestion
PERSONAL_ACCESS_TOKEN = demisto.params().get("personal_access_token", {}).get("password")
if not ((USERNAME and PASSWORD) or PERSONAL_ACCESS_TOKEN):
    return_error("You must provide either both USERNAME and PASSWORD, or a PERSONAL_ACCESS_TOKEN.")
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
 VERIFY_CERTIFICATE = not demisto.params().get("unsecure", False)
 
 # Support Credentials
 USERNAME = demisto.params()["credentials"]["identifier"]
 PASSWORD = demisto.params()["credentials"]["password"]
```
```suggestion
USERNAME = demisto.params().get("credentials", {}).get("identifier")
PASSWORD = demisto.params().get("credentials", {}).get("password")
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
+    "Accept": "application/json",
+    "X-Atlassian-Token": "no-check",
+}
+if PAT:
```
```suggestion
if PERSONAL_ACCESS_TOKEN:
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
-    except requests.exceptions.RequestException:  # This is the correct syntax
-        return_error(f"Failed to connect to - {full_url} - Please check the URL")
+def http_request(
+    method, full_url, data=None, params=None, is_test=False, is_download=False
```
```suggestion
    method, full_url, data=None, params=None, is_test=False, resp_type: str = "json",
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
 
     if is_test:
         return res
+    if is_download:
+        return res.content
     try:
         return res.json()
```
```suggestion
    try:
        if resp_type == "json":
            return res.json()
        elif resp_type == "content":
            return res.content
        else:
            return res
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -205,6 +261,15 @@ def get_content(key, title):
+    params = {"pageId": pageid}
+
+    full_url = SERVER + "/spaces/flyingpdf/pdfpageexport.action"
+    res = http_request("GET", full_url, None, params, is_download=True)
```
```suggestion
    res = http_request("GET", full_url, None, params, resp_type="content")
```

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.py:
```diff
@@ -15,50 +15,90 @@
+def http_request(
+    method, full_url, data=None, params=None, is_test=False, is_download=False
+):  # pragma: no cover
+    if PAT:
+        try:
+            res = requests.request(
+                method,
+                full_url,
+                verify=VERIFY_CERTIFICATE,
+                data=data,
+                headers=HEADERS,
+                params=params,
+            )
+        except requests.exceptions.RequestException:  # This is the correct syntax
+            return_error(f"Failed to connect to - {full_url} - Please check the URL")
+    else:
+        try:
+            res = requests.request(
+                method,
+                full_url,
+                verify=VERIFY_CERTIFICATE,
+                auth=(USERNAME, PASSWORD),
+                data=data,
+                headers=HEADERS,
+                params=params,
+            )
+        except requests.exceptions.RequestException:  # This is the correct syntax
+            return_error(f"Failed to connect to - {full_url} - Please check the URL")
```
```suggestion
    try:
        res = requests.request(
            method,
            full_url,
            verify=VERIFY_CERTIFICATE,
            auth=(USERNAME, PASSWORD) if not PERSONAL_ACCESS_TOKEN else None,
            data=data,
            headers=HEADERS,
            params=params,
        )
    except requests.exceptions.RequestException:  # This is the correct syntax
        return_error(f"Failed to connect to - {full_url} - Please check the URL")
```

mmhw reviewed (commented):
Hi @WildDogOne,
Thank you for your contribution!
Good work :)

Please pay attention to my comments.

Please feel free to reach out to me with any questions - I'm available here or on slack :)
Thanks again

WildDogOne commented:
Hello @mmhw  thanks for your code review.
I have added them to the pull request and did some tests on our confluence with the changes.
I added a few tweaks to get everything in order.

There is one last question remaining I guess, I think I have seen integrations output the file to the war room in a manner that someone could easily download them. For our usecase this is not necessary, but for the sake of completeness I'd like to add that.
Do you have any hints on how to do this?

At the moment the outputs look like this:
![image](https://github.com/user-attachments/assets/3c0429f7-c99c-424e-9f56-41522adcba56)

mmhw commented:
Hi @WildDogOne,
Regarding your question, returning `fileResult` in `demisto.results` should return an entry with the details about the file.
Try running the command in the Playground and see if the image actually downloads.
![image](https://github.com/user-attachments/assets/8af87017-e434-4170-90bb-7804d45c8b34)

Note that there are some comments of mine that you did not address.
Thanks again!

WildDogOne commented:
@mmhw I have cleaned up the rest of your commits, they somehow got hidden by github.
however I did not change the pageid to page_id in the confluence.yml since the other commands also use pageid.
I also am not happy with the implementation of the personal access token, because the UI does not reflect what the second password field is, and people will just have to know or assume that it might be the PAT

![image](https://github.com/user-attachments/assets/cd137ff1-a8cd-4109-8474-629023e5b914)

mmhw commented on Packs/Confluence/Integrations/Confluence/Confluence.yml:
```diff
@@ -15,7 +15,12 @@ configuration:
   type: 9
-  required: true
+  required: false
+- display: Personal Access Token
```
```suggestion
- display: ""
   displaypassword: Personal Access Token
```

mmhw commented:
> I have cleaned up the rest of your commits, they somehow got hidden by github.
> however I did not change the pageid to page_id in the confluence.yml since the other commands also use pageid.
> I also am not happy with the implementation of the personal access token, because the UI does not reflect what the second password field is, and people will just have to know or assume that it might be the PAT
You are right, we need to add `displaypassword: Personal Access Token`.
Additionally, it is also worth adding an explanation of how to obtain the token to the `Packs/Confluence/Integrations/Confluence/Confluence_description.md` file.
Thanks!

WildDogOne commented:
Alright, I had a look at the readme and noticed that the markdown version was empty.
I have now to the best of my ability converted the html readme to markdown and added the new functions (PAT and downloading PDF)

mmhw commented:
> Alright, I had a look at the readme and noticed that the markdown version was empty. I have now to the best of my ability converted the html readme to markdown and added the new functions (PAT and downloading PDF) edit: I was not able to add the new command to the TOC in the readme.md, since I don't know the UUID of the href. No idea how that stuff is generated

Note that the file `Confluence_description.md` appears as help for configuring the instance, so it should contain information about the arguments or how to obtain a token, etc.
<img width="1007" alt="image" src="https://github.com/user-attachments/assets/e7057c3b-5628-4411-82aa-7b5803e36624" />

The README.md file contains information about the integration and all the commands, etc.
You can generate the README file with the`demisto-sdk generate-docs -i Packs/Confluence/Integrations/Confluence/Confluence.yml` command, if you can't handle it, I'll do it.

mmhw commented:
Hi @WildDogOne, 
If you could please send me a video that shows the new command how it works, and that actually downloads a pdf file and you see it in WarRoom and Context Data.

Thanks!

WildDogOne commented:
I have attached a video of it in action.
https://github.com/user-attachments/assets/915cc4a3-05df-4d9e-ac40-21f28a89641f
Is there a less public way to share videos for future pull requests? I am building at least one for TOPdesk and Infoblox sooner or later.

I also tried to generate the readme, one error was related to not having a command example file, so I added that.
But even then I get errors that I am not 100% certain of what to do with them:

```
XXX@XXXone:/mnt$ demisto-sdk generate-docs -i Packs/Confluence/Integrations/Confluence/Confluence.yml -e Packs/Confluence/Integrations/Confluence/command_examples.txt 
Generating integration documentation
the commands example input is Packs/Confluence/Integrations/Confluence/command_examples.txt
found the following commands:
!confluence-create-space name=test description="testing space" key=TEST
!confluence-create-content space=DemistoContent title="test confluence integration" type=page body=testing
!confluence-list-spaces
!confluence-get-content key=DemistoContent title="test confluence integration"
!confluence-delete-content id=172723162
!confluence-update-content type=page pageid=172723162 currentversion=2 space=DemistoContent title="test confluence integration" body="new body"
!confluence-search-content cql="title=\"test confluence integration\""
!confluence-get-page-as-pdf pageid="123456"
Found existing integration, updating documentation...
        ⚙ Integration configuration has changed, updating...
Output file was saved to '/mnt/Packs/Confluence/Integrations/Confluence/README.md'
Found 10 possible errors:
        - The provided example for cmd !confluence-create-space name=test description="testing space" key=TEST has failed...
        - The provided example for cmd !confluence-create-content space=DemistoContent title="test confluence integration" type=page body=testing has failed...
        - The provided example for cmd !confluence-list-spaces has failed...
        - The provided example for cmd !confluence-get-content key=DemistoContent title="test confluence integration" has failed...
        - The provided example for cmd !confluence-delete-content id=172723162 has failed...
        - The provided example for cmd !confluence-update-content type=page pageid=172723162 currentversion=2 space=DemistoContent title="test confluence integration" body="new body" has failed...
        - The provided example for cmd !confluence-search-content cql="title=\"test confluence integration\"" has failed...
        - The provided example for cmd !confluence-get-page-as-pdf pageid="123456" has failed...
        - Unable to find configuration section line in README: Parameters table not found or incomplete in the README.
        - Skipping adding command 'confluence-get-page-as-pdf' as it was detected as a renamed from 'confluence-delete-content'

```

mmhw reviewed (approved):
Hi @WildDogOne,
I will merge this and take care of the README file.
Thanks for your work.
In the future, you can share videos with us either by email or via DFIR Slackץ

Thanks
